File: infrastructure/typescript/lib/database-stack.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import { RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib'
import { AttributeType, BillingMode, Table, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

export class DatabaseStack extends Stack {

  readonly messagesTable: Table;
  readonly channelsTable: Table;
  readonly connectionsTable: Table;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    this.connectionsTable = new Table(this, 'Connections', {
      partitionKey: { name: 'connectionId', type: AttributeType.STRING },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY, // NOT recommended for production use
      encryption: TableEncryption.AWS_MANAGED,
      pointInTimeRecovery: false // set to "true" to enable PITR
    });

    this.channelsTable = new Table(this, 'serverless-chat-channels', {
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      tableName: 'serverless-chat-channels',
      removalPolicy: RemovalPolicy.DESTROY, // NOT recommended for production use
      encryption: TableEncryption.AWS_MANAGED,
      pointInTimeRecovery: false // set to "true" to enable PITR
    });

    this.messagesTable = new Table(this, 'serverless-chat-messages', {
      partitionKey: {
        name: 'channelId',
        type: AttributeType.STRING
      },
      sortKey: {
        name: 'sentAt',
        type: AttributeType.STRING
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      tableName: 'serverless-chat-messages',
      removalPolicy: RemovalPolicy.DESTROY, // NOT recommended for production use
      encryption: TableEncryption.AWS_MANAGED,
      pointInTimeRecovery: false // set to "true" to enable PITR
    });
  }
};

File: infrastructure-net7/src/Infrastructure/Stacks/DatabaseStack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.DynamoDB;
using Constructs;

namespace Infrastructure.Stacks
{
    public class DatabaseStack : Stack
    {
        public Table MessagesTable { get; private set; }
        public Table ChannelsTable { get; private set; }
        public Table ConnectionsTable { get; private set; }

        internal DatabaseStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
        {
            this.ConnectionsTable = new Table(this, "Connections", new TableProps()
            {
                PartitionKey = new Attribute() { Name = "connectionId", Type = AttributeType.STRING },
                BillingMode = BillingMode.PAY_PER_REQUEST,
                RemovalPolicy = RemovalPolicy.DESTROY, // NOT recommended for production use
                Encryption = TableEncryption.AWS_MANAGED,
                PointInTimeRecovery = false // set "true" to enable PITR
            });
            
            this.ChannelsTable = new Table(this, "serverless-chat-channels", new TableProps()
            {
                PartitionKey = new Attribute() { Name = "id", Type = AttributeType.STRING },
                TableName = "serverless-chat-channels",
                BillingMode = BillingMode.PAY_PER_REQUEST,
                RemovalPolicy = RemovalPolicy.DESTROY, // NOT recommended for production use
                Encryption = TableEncryption.AWS_MANAGED,
                PointInTimeRecovery = false // set "true" to enable PITR
            });
            
            this.MessagesTable = new Table(this, "serverless-chat-messages", new TableProps()
            {
                PartitionKey = new Attribute() { Name = "channelId", Type = AttributeType.STRING },
                SortKey = new Attribute() { Name = "sentAt", Type = AttributeType.STRING },
                BillingMode = BillingMode.PAY_PER_REQUEST,
                TableName = "serverless-chat-messages",
                RemovalPolicy = RemovalPolicy.DESTROY, // NOT recommended for production use
                Encryption = TableEncryption.AWS_MANAGED,
                PointInTimeRecovery = false // set "true" to enable PITR
            });
        }
    }
}
